### PR TITLE
Support optimization backends via package extensions

### PR DESCRIPTION
**Pull request by alyst**
This is another set of changes from #193 and my staging branch:

* `optimizer` is removed from the `Sem` object. `Sem` defines the SEM model, which is independent from how the parameters of this model are fit.
   Now the optimizer has to be specified for `sem_fit()` call.
   *(I guess the optimizer was a part of `Sem`, because some optimizers allow specifying parameter constraints, which is the part of the model.
    But that constraints specification was optimizer-dependent. In future, it should be possible to implement optimizer-agnostic constraints specification.)*
    That change simplifies the code a bit, since `Sem.optimizer` was never really used except to call `sem_fit()`.
    Plus, it allows fitting the same model with different optimizer backends without recreating the `Sem` object.
* Added support for different optimization backends via package extension mechanisms.
   Implementation details:
   * `SemOptimizer{E}`: engine type parameter `E` was added to the abstract `SemOptimizer` class,
     i.e. `SemOptimizer{E}(args...; kwargs...)` constructs the optimizer object for the `E` backend (e.g. `E = :Optim` for *Optim.jl*).
   * To add support for a given backend, one needs to implement:
     1. `SemOptimizer{E}(...)` constructor, e.g.  `SemOptimizer{:Optim}(...)` creates `SemOptimizerOptim <: SemOptimizer{:Optim}` object;
     2. `sem_fit(optim::O, model, ...) where O <: SemOptimizer{E}` call
   * There is also a call `sem_fit(model; engine, kwargs...)`, which first creates `SemOptimizer{engine}(;kwargs...)` optimizer and then calls its `sem_fit()`
   * *Optim.jl* is the default backend (`E = :Optim`), so it remains a hard dependency of *SEM.jl*
   * *NLopt.jl* support (`E = :NLopt`) is moved into the `SEMNLOptExt` extension (it gets automatically activated once the *NLopt* package is loaded into an activate session)
   *  Converted the [ProximalSEM.jl](https://github.com/StructuralEquationModels/ProximalSEM.jl) code into the `SEMProximalOptExt` extension (`E = :Proximal`) to enable the [ProximalAlgorithms.jl](https://github.com/JuliaFirstOrder/ProximalAlgorithms.jl) backend support.
 * Added support for specifying lower and upper bounds of model parameters when calling `sem_fit()` for the *Optim.jl* backend.

### CODE REVIEW

**Comment by alyst** [comment]
@Maximilian-Stefan-Ernst Also, I haven't touched the sources layout, but you may consider merging the optimizer parts in `diff` folder (`SemOptimizer`-derived type) into the `optimizer` folder (`sem_fit()` implementation for that type).
If you think that's a good idea, I can merge the folders  in this PR.

**Comment by Maximilian-Stefan-Ernst** [comment]
I think that's definitely a good idea, the "diff"-folder is a misnomer anyways^^

File: src/additional_functions/start_val/start_partable.jl [hunk 1]
```diff
@@ -3,39 +3,22 @@
...
-end
-
-function start_parameter_table(ram::RAMMatrices; partable::ParameterTable, kwargs...)
-    start_val = zeros(0)
-
-    param_indices = Dict(param => i for (i, param) in enumerate(params(ram)))
-    start_col = partable.columns[:start]
+function start_parameter_table(partable::ParameterTable)
+    start_vals = zeros(eltype(partable.columns[:start]), nparams(partable))
+    param_indices = Dict(param => i for (i, param) in enumerate(params(partable)))
```
Comment (Maximilian-Stefan-Ernst) [inline src/additional_functions/start_val/start_partable.jl:11]: I agree that my code was kind of messy and rewriting is good, one thing that I would maybe keep though is checking whether the parameter table and the model have the same parameters / taking the param_indices from the model, not the table. Because I want to avoid a situation, where someone creates a model from a table, then maybe changes the table / creates a new one where the parameter indices are in a different order, and then gets wrong starting values.

Comment (alyst) [reply src/additional_functions/start_val/start_partable.jl:11]: Oh, that's clearly a bug on my side, the indices should be from `ram`. Thanks for noticing!

File: src/optimizer/documentation.jl [hunk 2]
```diff
@@ -1,29 +1,122 @@
...
+            def = in(p, varparams) ? variance_default : default
+            isnothing(bounds) ? def : get(bounds, p, def)
+        end for p in SEM.params(model)
+    ]
+
+    return res
+end
+
+# define a vector of parameter upper bounds: use user-specified vector as is
+function upper_bounds(bounds::AbstractVector, model::AbstractSem; default::Number)
```
Comment (Maximilian-Stefan-Ernst) [inline src/optimizer/documentation.jl:99]: the argument `default` is not used in the function

File: src/optimizer/documentation.jl [hunk 3]
```diff
@@ -1,29 +1,122 @@
...
+    @assert length(start_val) == nparams(model)
+    return start_val
+end
+
+# define a vector of parameter lower bounds: use user-specified vector as is
+function lower_bounds(
+    bounds::AbstractVector,
+    model::AbstractSem;
+    default::Number,
+    variance_default::Number,
```
Comment (Maximilian-Stefan-Ernst) [inline src/optimizer/documentation.jl:70]: These two arguments are not used in the function

File: src/optimizer/documentation.jl [hunk 4]
```diff
@@ -1,29 +1,122 @@
 """
```
Comment (Maximilian-Stefan-Ernst) [inline src/optimizer/documentation.jl:1]: I think the file "documentation.jl" should now be renamed, since it also implements general functionality like starting values and bounds - so maybe we should call it something like "optimizer.jl" or "sem_fit.jl".

File: src/types.jl [hunk 5]
```diff
@@ -105,73 +116,66 @@ abstract type SemImply end
 abstract type SemImplySymbolic <: SemImply end
 
 """
-    Sem(;observed = SemObservedData, imply = RAM, loss = SemML, optimizer = SemOptimizerOptim, kwargs...)
+    Sem(;observed = SemObservedData, imply = RAM, loss = SemML, kwargs...)
 
 Constructor for the basic `Sem` type.
-All additional kwargs are passed down to the constructors for the observed, imply, loss and optimizer fields.
+All additional kwargs are passed down to the constructors for the observed, imply, loss fields.
```
Comment (Maximilian-Stefan-Ernst) [inline src/types.jl:122]:
```suggestion
All additional kwargs are passed down to the constructors for the observed, imply, and loss fields.
```

File: ext/SEMProximalOptExt.jl [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+using StructuralEquationModels
+using ProximalCore, ProximalAlgorithms, ProximalOperators
+
+export SemOptimizerProximal
+
+SEM = StructuralEquationModels
+
+#ProximalCore.prox!(y, f, x, gamma) = ProximalOperators.prox!(y, f, x, gamma)
+
+include("optimizer/ProximalAlgorithms.jl")
```
Comment (Maximilian-Stefan-Ernst) [inline ext/SEMProximalOptExt.jl:12]: Should we maybe just add the code from "optimizer/ProximalAlgorithms.jl" here? So we only have one file per extension atm.

**Comment by Maximilian-Stefan-Ernst** [comment]
I finally got to the review, thanks for the great changes (again)! I've made some minor comments, I think the only thing that would be left before merging would be to move the tests from the [ProximalSEM.jl](https://github.com/StructuralEquationModels/ProximalSEM.jl) package to the tests of this package.

File: test/examples/political_democracy/constructor.jl [hunk 7]
```diff
@@ -1,10 +1,12 @@
...
-using Random
+using Random, NLopt
 
 ############################################################################################
 ### models w.o. meanstructure
 ############################################################################################
 
+semoptimizer = SemOptimizer(engine = opt_engine)
+
 model_ml = Sem(specification = spec, data = dat, optimizer = semoptimizer)
```
Comment (Maximilian-Stefan-Ernst) [inline test/examples/political_democracy/constructor.jl:10]: I'm not sure this works - since the optimizer now lives outside of the Sem object, the keyword `optimizer` will just be ignored, and the optimizer has to be added to each call of sem_fit instead, right?

**Comment by Maximilian-Stefan-Ernst** [comment]
I moved the tests from ProximalSEM.

**Comment by alyst** [comment]
@Maximilian-Stefan-Ernst Thank you for a thorough review and the fixes! I will address the remaining ones in 1-2 days.

BTW, have you tried/considered using [Manopt.jl](https://manoptjl.org/stable/) for SEM optimization?
With [Manifolds.jl](https://juliamanifolds.github.io/Manifolds.jl/stable/) it should be possible to construct a combined manifold that restricts S to SPD.
Anyway, with this PR it should be easier to try.

File: src/optimizer/documentation.jl [hunk 8]
```diff
@@ -1,29 +1,122 @@
...
+        ),
+    )
+    return bounds
+end
+
+# define a vector of parameter lower bounds given a dictionary and default values
+function lower_bounds(
+    bounds::Union{AbstractDict, Nothing},
+    model::AbstractSem;
+    default::Number,
```
[the remark this answers is not in the document]
Comment (alyst) [reply src/optimizer/documentation.jl]: I refactored bounds part -- now there is a single `prepare_param_bounds()` function (to reduce code duplication), which addresses some of your "unused" comments.
I did not put default value intentionally -- the public `sem_fit(::SemOptimizerOptim, ...)` has the kwargs with the default values, and it passes them to this function.
So if anything changes in the `sem_fit()`, which is not compatible with the `xxx_bounds()` function -- it should be easier to detect.
I have put the comment about the unusedness (the parameters are kept so that the function signature stays the same).

File: src/additional_functions/start_val/start_partable.jl (hunk 1 above)
Comment (alyst) [reply src/additional_functions/start_val/start_partable.jl]: By looking at the logic further, I realized that it could be simplified: instead of `sem_fit(..., start_val = start_parameter_table, parameter_table = partable)` it is easier to just support `sem_fit(..., start_val = partable)`.
So `prepare_start_params()` is split into several methods instead of internal type-dependent `if`.
And `start_parameter_table()` is refactored into `prepare_start_params(start_val::ParameterTable, ...)`.

File: src/optimizer/documentation.jl (hunk 4 above)
Comment (alyst) [reply src/optimizer/documentation.jl]: I agree. I renamed it into the `abstract.jl`.
There are already `abstract.jl` files in the other source code branches, so it follows that pattern.
I think in general all the remaining `documentation.jl` could be renamed that way.

File: ext/SEMProximalOptExt.jl (hunk 6 above)
Comment (alyst) [reply ext/SEMProximalOptExt.jl:12]: I think it's nice to follow the same pattern as for the main package -- one lean main source file, which has all the `using` and includes all the source files.
Plus, some backends may have more code and/or more source files, and that rule makes the extensions structure more consistent.
We can remove the `optimizer` folder, though.

File: test/examples/political_democracy/constructor.jl (hunk 7 above)
Comment (alyst) [reply test/examples/political_democracy/constructor.jl]: No, it doesn't work -- I forgot to update it. Should be fixed now.
That, sort of, highlights one limitation of the current API: it's based on passing arguments via keywords through the hierarchy of ctor calls.
And non-matching keywords are silently ignored, because each ctor cannot know whether it's an incorrect keyword or it is the correct keyword for another ctor.
So the bug like this was not automatically detected by the tests.

**Comment by alyst** [comment]
@Maximilian-Stefan-Ernst I think I have addressed all of your comments, thanks again for the review!
Let me know if there are still things to tweak. If you think it's ready, also let me know -- I will squash the fixup commits to cleanup the history before merging it.

File: src/optimizer/documentation.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (Maximilian-Stefan-Ernst) [reply src/optimizer/documentation.jl]: Perfect, thanks.

File: ext/SEMProximalOptExt.jl (hunk 6 above)
Comment (Maximilian-Stefan-Ernst) [reply ext/SEMProximalOptExt.jl:12]: Sounds good, then I would prefer to remove the optimizer folder, and have all files of each extension in separate folders, i.e. ext/SEMProximalOptExt/SEMProximalOptExt.jl.

File: test/examples/political_democracy/constructor.jl (hunk 7 above)
Comment (alyst) [reply test/examples/political_democracy/constructor.jl]: That is probably the topic for a separate issue, but I think the main ctors for `SemSpecification`, `SemObserved` and `Sem` that do all of the actual construction work should get the essential inputs as positional arguments (that is also important for type-dependent dispatch) and don't do keyword arguments passthrough.
Then it might be possible to have a simple single ctor-like function for `Sem()` that internally creates all 3 objects like it does now, but with more constrained set of keywords.

Also, explicitly creating `SemSpecification`, `SemObserved` and explicitly passing them to `Sem` might in the end be not so much more complex for the user.
The real design problem with that approach is that `SemSpec` and `SemObserved` are independent objects, but they are constrained to have the same set and same order of observed variables.
Maybe it is possible to utilize `ScopedValues` from 1.11, so that the user explicitly calls `Sem(SemSpec(...; kwargs1...), SemObserved(...; kwargs2); kwargs3...)`, but the observed variables order is implicitly passed to `SemSpec()` and `SemObserved()` via `ScopedValues`.

File: ext/SEMProximalOptExt.jl (hunk 6 above)
Comment (alyst) [reply ext/SEMProximalOptExt.jl]: That's what I was also thinking about. I have updated the layout.

**Comment by alyst** [comment]
I've cleaned up the commit history, so it should be ok to merge if no other changes are required.

File: test/examples/political_democracy/constructor.jl (hunk 7 above)
Comment (Maximilian-Stefan-Ernst) [reply test/examples/political_democracy/constructor.jl]: I references this in a issue for now, so we can revisit it at a later point 👍🏽

**Comment by Maximilian-Stefan-Ernst** [comment]
@alyst: W.r.t. Manopt.jl: The problem why I did not manage to implement the SPD constraint with ProximalAlgorithms.jl is because we have this situation where the model-implied covariance matrix is a function of the parameters, so we have parameters -> cov matrix -> loss, and we have to constrain in the middle instead of the parameter space. I am not sure how we could alleviate this with Manopt - I believe we would have to figure out the manifold of the parameter space?
